react/training-project/src/redux/Auth/actions.js

```diff
@@ -0,0 +1,74 @@
+import AuthService from '../../services/AuthService';
+
+export const actions = {
+  LOG_IN: '@@AUTH/LOG_IN',
+  LOG_IN_SUCCESS: '@@AUTH/LOG_IN_SUCCESS',
+  LOG_IN_FAIL: '@@AUTH/LOG_IN_FAIL',
+  LOG_IN_PROCESSING: '@AUTH/LOG_IN_PROCESSING',
+  LOG_OUT: '@@AUTH/V',
+  LOG_OUT_SUCCESS: '@@AUTH/LOG_OUT_SUCCESS',
+  LOG_OUT_FAIL: '@@AUTH/LOG_OUT_FAIL'
+};
+
+const privateActionsCreators = {
+  logInProcessing: bool => ({
+    type: actions.LOG_IN_PROCESSING,
+    payload: bool
+  }),
+
+  logInFail: (bool, processing) => ({
+    type: actions.LOG_IN_FAIL,
+    payload: {
+      bool,
+      processing
+    }
+  }),
+
+  logInSuccess: (token, processing, loginFail) => ({
+    type: actions.LOG_IN_SUCCESS,
+    payload: {
+      token,
+      processing,
+      loginFail
+    }
+  }),
+
+  logOutSuccess: () => ({
+    type: actions.LOG_OUT_SUCCESS,
+    payload: null
+  })
+};
+
+const actionCreators = {
+  logIn: ({ username, password }) => async dispatch => {
+    dispatch(privateActionsCreators.logInProcessing(true));
+    const response = await AuthService.logIn({ username, password });
+
+    if (response.ok && response.data.length > 0) {
```

Comment:
data is an array? :thinking: If it is, then, you should check the existence of `response.data`. I know that if `response.ok` is true, `response.data` should exists, but it's always better to check this things and prevent some errors.